### PR TITLE
Make application ID configurable in trigger service

### LINKED ISSUE #7029: Make application id configurable in DAML Script
This primarily matters if you use auth where the app id in your token must match what you use for command submissions.
We could consider making this configurable per participant.

### LINKED ISSUE #7671: Decide what application ID to assign in trigger service
Quoting https://github.com/digital-asset/daml/pull/7654/files#r503775725

> Not quite sure the behavior in the trigger service that we have atm is sensible. For the standalone trigger runner we made the application id configurable via a CLI parameter. Not quite sure what the best approach here is.
> 
>   1. We could mandate it as a parameter when starting the trigger but that seems a bit annoying
>   2. We could make it optional and default to the UUID of the trigger which could be reasonable
>   3. We could also try to infer it from the token but that gets a bit annoying since I’d prefer not to parse the token so the middleware would need to pass this on.
> 
> 
> Let’s leave this out of this PR but I think we need something more sensible than what we do now.

### PR DESCRIPTION
**Pull request by cocreature**
fixes #7671 

changelog_begin

- [Trigger Service] The application id used by a trigger can now be
  configured by an optional `applicationId` in the start request.

changelog_end

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

**Review by S11001001 — APPROVED** [review]

File: triggers/service/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Server.scala [hunk 1]
```diff
@@ -103,9 +103,11 @@ class Server(
   }
 
   private def restartTriggers(triggers: Vector[RunningTrigger]): Either[String, Unit] = {
-    import cats.implicits._ // needed for traverse
-    triggers.traverse_(t =>
-      startTrigger(t.triggerParty, t.triggerName, t.triggerToken, Some(t.triggerInstance)))
+    import cats.implicits._ // needed for traverse for
```
Comment (aherrmann-da) [inline triggers/service/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Server.scala:106]:
```suggestion
    import cats.implicits._ // needed for traverse
```

**Review by aherrmann-da — COMMENTED** [review]
LGTM. Though, this still leaves the auth aspect open.

Currently, the auth middleware cannot request a token for a particular application id. Meaning, we either need to issue tokens that leave the application id out, i.e. wildcard tokens in the application id, or users of the trigger service need to make sure to always request triggers with the one application id that matches the auth service. The trigger service tests currently use wildcard tokens, see [here](https://github.com/digital-asset/daml/blob/7c938130d76dfc867debadc36e508b01a6d3fb88/triggers/service/src/test/scala/com/digitalasset/daml/lf/engine/trigger/TriggerServiceFixture.scala#L184)

Making the application id configurable per trigger suggests that we should either just stick to wildcard tokens or we need to extend the auth middleware to support requesting a token for a specific application id. E.g. by adding a scope for it.

**Comment by cocreature** [comment]
> Currently, the auth middleware cannot request a token for a particular application id. Meaning, we either need to issue tokens that leave the application id out, i.e. wildcard tokens in the application id, or users of the trigger service need to make sure to always request triggers with the one application id that matches the auth service. The trigger service tests currently use wildcard tokens, see [here](https://github.com/digital-asset/daml/blob/7c938130d76dfc867debadc36e508b01a6d3fb88/triggers/service/src/test/scala/com/digitalasset/daml/lf/engine/trigger/TriggerServiceFixture.scala#L184)
> 
> Making the application id configurable per trigger suggests that we should either just stick to wildcard tokens or we need to extend the auth middleware to support requesting a token for a specific application id. E.g. by adding a scope for it.

Yeah that’s true. It doesn’t change this though. Previously, the only option was wildcards tokens since the application id was random. Adding it to the middleware seems sensible. I think the semantics should be such that if you specify a given application id you should either get back a token valid for the application id meaning either a wildcard or one that has the given application id. Are you fine with tackling this in a separate PR?

**Comment by aherrmann-da** [comment]
> I think the semantics should be such that if you specify a given application id you should either get back a token valid for the application id meaning either a wildcard or one that has the given application id.

Agreed. Ultimately, the authorization service decides what the token looks like. The middleware can only verify whether it meets the requested claims. Regarding the application id this check would either require a wildcard or a match.

> Are you fine with tackling this in a separate PR?

Yes, I'll open a ticket to track it.

**Review by aherrmann-da — APPROVED** [review]

**Comment by cocreature** [comment]
Opened #7978

**Pull request merged**